Calling configure on WebCodecs Video Encoder triggers a new key frame each time even if the configuration is only changing the bitrate

youennf opened the pull request:
#### 37dfb4fa3c73132a0d3b5d90aa0cb6253cad64bb
<pre>
Calling configure on WebCodecs Video Encoder triggers a new key frame each time even if the configuration is only changing the bitrate
<a href="https://bugs.webkit.org/show_bug.cgi?id=261402">https://bugs.webkit.org/show_bug.cgi?id=261402</a>
<a href="https://rdar.apple.com/115482127">rdar://115482127</a>

Reviewed by Jean-Yves Avenard.

Check the new configuration is the same as the new one, except for bitrate and framerate.
If that is the case, we call setRate instead of recreating the encoder.
Implement setRate for VPX and remote encoders.

Covered by newly added test.

* LayoutTests/http/wpt/webcodecs/encode-bitrate-change-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/encode-bitrate-change.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::updateRates):
(WebCore::WebCodecsVideoEncoder::configure):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoderConfig.h:
(WebCore::isSameConfigurationExceptBitrateAndFramerate):
* Source/WebCore/platform/VideoEncoder.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp:
(WebCore::LibWebRTCVPXVideoEncoder::setRates):
(WebCore::computeAllocation):
(WebCore::LibWebRTCVPXInternalVideoEncoder::initialize):
(WebCore::LibWebRTCVPXInternalVideoEncoder::encode):
(WebCore::LibWebRTCVPXInternalVideoEncoder::setRates):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoEncoder::setRates):
(WebKit::RemoteVideoEncoder::flush):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::setEncodeRatesCallback):
(WebKit::LibWebRTCCodecs::setEncodeRates):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:

Originally-landed-as: 4c9425f9126b. <a href="https://rdar.apple.com/115482127">rdar://115482127</a>
Canonical link: <a href="https://commits.webkit.org/280771@main">https://commits.webkit.org/280771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2838a5253fdec93e0e0b5852529d817616732e57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57578 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61200 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8023 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59706 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46621 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5689 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49742 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27487 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31400 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7040 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7026 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53357 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62880 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1491 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7405 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53881 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1498 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53986 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12736 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1268 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32735 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33820 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33566 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
```diff
@@ -113,6 +113,28 @@ static ExceptionOr<VideoEncoder::Config> createVideoEncoderConfig(const WebCodec
+    auto framerate = config.framerate.value_or(0);
+
+    m_isMessageQueueBlocked = true;
+    m_internalEncoder->setRates(bitrate, framerate, [this, weakedThis = WeakPtr { *this }, bitrate, framerate]() mutable {
```
`weakThis`

youennf replied:
done

jyavenard commented on Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp:
```diff
@@ -75,8 +79,7 @@ class LibWebRTCVPXInternalVideoEncoder : public ThreadSafeRefCounted<LibWebRTCVP
     int64_t m_timestamp { 0 };
     std::optional<uint64_t> m_duration;
     bool m_isClosed { false };
-    uint64_t m_width { 0 };
```
that seems an odd data type choice for resolution.

Are there any doing more than even 16 bits? (16384x16384)

youennf replied:
JS allows uint32_t, uint64_t was a mistake that we correct by storing the Config.

jyavenard commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
```diff
@@ -292,12 +293,17 @@ void RemoteVideoEncoder::encode(RawFrame&& rawFrame, bool shouldGenerateKeyFrame
 
+void RemoteVideoEncoder::setRates(uint64_t bitRate, double frameRate, Function<void()>&& callback)
+{
+    auto bitRateInKbps = bitRate / 1000;
```
K is kelvin
k is kilo

youennf replied:
In the code base, I can see a lot of Kbps, no KBPS, and _kbps.
I think this is known here.

youennf committed:
Calling configure on WebCodecs Video Encoder triggers a new key frame each time even if the configuration is only changing the bitrate
https://bugs.webkit.org/show_bug.cgi?id=261402
rdar://115482127

Reviewed by Jean-Yves Avenard.

Check the new configuration is the same as the new one, except for bitrate and framerate.
If that is the case, we call setRate instead of recreating the encoder.
Implement setRate for VPX and remote encoders.

Covered by newly added test.

* LayoutTests/http/wpt/webcodecs/encode-bitrate-change-expected.txt: Added.
* LayoutTests/http/wpt/webcodecs/encode-bitrate-change.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::updateRates):
(WebCore::WebCodecsVideoEncoder::configure):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoderConfig.h:
(WebCore::isSameConfigurationExceptBitrateAndFramerate):
* Source/WebCore/platform/VideoEncoder.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp:
(WebCore::LibWebRTCVPXVideoEncoder::setRates):
(WebCore::computeAllocation):
(WebCore::LibWebRTCVPXInternalVideoEncoder::initialize):
(WebCore::LibWebRTCVPXInternalVideoEncoder::encode):
(WebCore::LibWebRTCVPXInternalVideoEncoder::setRates):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoEncoder::setRates):
(WebKit::RemoteVideoEncoder::flush):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::setEncodeRatesCallback):
(WebKit::LibWebRTCCodecs::setEncodeRates):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:

Originally-landed-as: 4c9425f9126b. rdar://115482127
Canonical link: https://commits.webkit.org/280771@main
